tests/run-make/rustdoc-io-error/rmake.rs

```diff
@@ -0,0 +1,41 @@
+// This test verifies that rustdoc doesn't ICE when it encounters an IO error
+// while generating files. Ideally this would be a rustdoc-ui test, so we could
+// verify the error message as well.
+//
+// It operates by creating a temporary directory and modifying its
+// permissions so that it is not writable. We have to take special care to set
+// the permissions back to normal so that it's able to be deleted later.
+
+use run_make_support::{rustdoc, tmp_dir};
+use std::fs;
+
+fn main() {
+    // FIXME: Remove this line once figured out what's wrong.
+    let out_dir = tmp_dir().join("rustdoc-io-error");
+    let output = fs::create_dir(&out_dir).unwrap();
+    let mut permissions = fs::metadata(&out_dir).unwrap().permissions();
+    let original_permissions = permissions.clone();
+
+    permissions.set_readonly(true);
+    fs::set_permissions(&out_dir, permissions).unwrap();
```

Comment:
Docs for [`Permissions::set_readonly`](https://doc.rust-lang.org/std/fs/struct.Permissions.html#method.set_readonly) says that this "makes the file world-writable on Unix. You can use the [PermissionsExt](https://doc.rust-lang.org/std/os/unix/fs/trait.PermissionsExt.html) trait on Unix to avoid this issue."

Maybe try using `permissions.set_mode(0o444)` (read-only for everyone) then later `permissions.set_mode(0o666)` (read-write for everyone)?